barretenberg/cpp/src/barretenberg/ecc/scalar_multiplication/sorted_msm.test.cpp

```diff
@@ -0,0 +1,254 @@
+{
+    using Curve = TypeParam;
+    using G1 = typename Curve::AffineElement;
+    using Sorter = MsmSorter<Curve>;
+    using AdditionSequences = typename Sorter::AdditionSequences;
+
+    // Generate random MSM inputs based on a set of sequence counts
+    std::array<uint64_t, 3> sequence_counts{ 5, 2, 3 };
+    auto [num_points, points, scalars, msm_result] =
+        TestFixture::generate_test_data_from_sequence_counts(sequence_counts);
+
+    AdditionSequences addition_sequences{ sequence_counts, points, {} };
+
+    std::vector<G1> expected_points;
+    size_t point_idx = 0;
+    for (auto count : sequence_counts) {
+        G1 sum = points[point_idx++];
+        for (size_t i = 1; i < count; ++i) {
+            sum = sum + points[point_idx++];
+        }
+        expected_points.emplace_back(sum);
+    }
+
+    Sorter msm_sorter(num_points);
+    msm_sorter.batched_affine_add_in_place(addition_sequences);
+
+    for (size_t idx = 0; idx < expected_points.size(); ++idx) {
+        EXPECT_EQ(expected_points[idx], points[idx]);
+    }
+}
+
+// Test generation of point addition sequences from an arbitrary set of points and scalars
+TYPED_TEST(SortedMsmTests, GenerateAdditionSequences)
+{
+    using Curve = TypeParam;
+    using G1 = typename Curve::AffineElement;
+    using Fr = typename Curve::ScalarField;
+    using Sorter = MsmSorter<Curve>;
+    using AdditionSequences = typename Sorter::AdditionSequences;
+
+    // Generate random MSM inputs based on a set of sequence counts
+    std::array<uint64_t, 3> sequence_counts{ 5, 2, 3 };
+    auto [num_points, points, scalars, expected_msm_result] =
+        TestFixture::generate_test_data_from_sequence_counts(sequence_counts);
+
+    Sorter msm_sorter{ num_points };
+    AdditionSequences result = msm_sorter.construct_addition_sequences(scalars, points);
+
+    // The resulting sequence counts should match expectation but only as multisets
+    std::multiset<Fr> expected_sequence_counts(sequence_counts.begin(), sequence_counts.end());
+    std::multiset<Fr> result_sequence_counts(result.sequence_counts.begin(), result.sequence_counts.end());
+    EXPECT_EQ(expected_sequence_counts, result_sequence_counts);
+
+    // The result points will be sorted but should match the original as multisets
+    std::multiset<G1> expected_points(points.begin(), points.end());
+    std::multiset<G1> result_points(result.points.begin(), result.points.end());
+    EXPECT_EQ(expected_points, result_points);
+
+    G1 msm_result;
+    msm_result.self_set_infinity();
+    size_t scalar_idx = 0;
+    size_t point_idx = 0;
+    for (auto count : result.sequence_counts) {
+        for (size_t i = 0; i < count; ++i) {
+            msm_result = msm_result + result.points[point_idx] * msm_sorter.unique_scalars[scalar_idx];
+            point_idx++;
+        }
+        scalar_idx++;
+    }
+
+    EXPECT_EQ(msm_result, expected_msm_result);
+}
+
+TYPED_TEST(SortedMsmTests, ReduceMsmInputsSimple)
```

Comment:
Could you add a comment what this test is testing?

Reply:
done, thanks, not sure why I missed this